### ISSUE TITLE
KeywordRecognizer prevents system from sleeping automatically

### ISSUE THREAD

**Issue by assitive** [Azure-Samples/cognitive-services-speech-sdk#761]
**Describe the bug**
Using C# KeywordRecognizer class prevents Windows 10 from going to sleep automatically.

**To Reproduce**
1. In System's Power & Sleep dialog, change screen off time to Never and Sleep time to 1 minute.
2. Create a keyword model file from MS keyword portal
3. Create a C# problem with the following line of code:
new KeywordRecognizer(AudioConfig.FromDefaultMicrophoneInput()).RecognizeOnceAsync(KeywordRecognitionModel.FromFile("test.table")).Wait();
4. Build and run the program
5. Wait for one minute or more without using the computer, system never goes to sleep. However if I exit the test program, then system goes to sleep after one minute. 

**Expected behavior**
Wait for one minute, Windows 10 should automatically go to sleep.

**Version of the Cognitive Services Speech SDK**
microsoft.cognitiveservices.speech 1.12.0

**Platform, Operating System, and Programming Language**
 - Windows 10
 - x64
 - C#
 - USB wireless headset

**Additional context**
 - "powercfg -requests" in CMD shows:
SYSTEM:
[DRIVER] USB Audio Device (USB\VID_0B0E&PID_1022&MI_00\6&257614f2&0&0000)
An audio stream is currently in use.
[DRIVER] Legacy Kernel Caller

**Comment by trrwilson** [comment]
Hello! Thanks for getting in touch.

There are two important elements to this:
* It's *by design* that active audio capture stops the OS from going to sleep. You can imagine how frustrating it would be if you were on an involved call for a few minutes and then you were suddenly cut off because you passed the sleep threshold; to avoid that, ongoing audio capture is treated as an activity signal just like moving your mouse or striking your keyboard. The keyword spotter is "just another consumer of audio capture," which means that the OS is blocked from going to a low power state while it's listening for a keyword.
* Applications *should*, however, be able to get process lifetime signals (e.g. Suspending calls in UWP) when the OS would like to go to sleep and respond to these by stopping the capture (in this case, via the stop API on the recognizer). This gives a nice layer of control to an application that can decide if it should continue blocking sleep or stop things.

Additionally, the screen off signal is often useful in deciding when audio capture should be stopped. Not if you want to be able to wake things up via voice, of course, but if it's an experience that the user will expect the screen to be on for, that can be a great help in keeping sleep transitions clear.

What kind of application (Win32, UWP, etc.) are you involved with? We may be able to advise on reference examples of how to register and respond to the lifecycle notifications, which will be the best bet.

**Comment by assitive** [comment]
Thanks for the response. We are developing a classic Win32 desktop application with C#. We are expecting this to work similarly to Cortana which doesn't prevent system from going to sleep. And we do NOT intend to give our customers the ability to wake up system or our program in sleeping mode. The second point sounds interesting. How can we get the process lifetime signal in our program? If we do get the signal and stop keyword recognition, does the system go to sleep shortly after this signal?

**Comment by trrwilson** [comment]
Thank you for the additional detail.

One point to clarify is that Cortana has the same limitations unless the device supports a hardware keyword spotter (low-power and independent of the OS audio stack for its activation): having keyword activation on will prevent the computer from going completely to sleep with the same driver output in `powercfg -requests`.

The good news is that you have a lot of great tools available in a Win32 application. Talking with some experts on power management, the first recommendation I got was to monitor for screen off events and to use that as the signal to stop keyword recognition--you can further caveat this by not turning it off if the system isn't configured to sleep if it's on AC power, but that's icing on the cake.

[PowerSettingRegisterNotification](https://docs.microsoft.com/en-us/windows/win32/api/powersetting/nf-powersetting-powersettingregisternotification) is the low-level API that monitors for the display (and many other) events. It doesn't have a direct C# version, but it's fully accessible via interop and many wrappers have already been written for it by the community. For one example (note: not affiliated with Microsoft or GitHub, external site, etc.): http://www.pinvoke.net/default.aspx/powrprof/PowerSettingRegisterNotification.html

To summarize:
* The app registers for a `GUID_CONSOLE_DISPLAY_STATE` notification via `PowerSettingRegisterNotification` (C# interop)
* Upon receiving the event,
  * (optional) Check if the device is on AC power and configured to not sleep on AC power--if both true, done (let it run)
  * Call `StopKeywordRecognitionAsync` to end the ongoing recognition and stop the use of the input audio resource
* Once active audio streams are stopped, I believe the countdown *starts* -- e.g. if the screen turns off after 10 minutes and the computer is configured to sleep after 30 minutes of inactivity, the net would be 40 minutes since "inactivity" technically only started once the stream was shut off
* Upon the screen turning back on, `StartKeywordRecognitionAsync` again to resume

**Comment by assitive** [comment]
The suggestion does not work for us. It's not acceptable user experience the feature would stop working for the extra 40 minutes after receiving the event. I understand the argument that system should not go to sleep when one is on a call. But using mic for wake word listening is another thing. People doesn't get frustrated if system goes to sleep while the system is listening for wake word. There is simply no interaction between the user and the computer when it happens.

**Comment by assitive** [comment]
By the way Travis, it sounded to me this is limited by hardware's power management capability. Can you suggest a few microphones with a keyword spotter? If we use one of those mic's, is there any code change needed in order to support keyword listening and auto sleeping?

**Comment by trrwilson** [comment]
I certainly understand the frustration and we share it when we're creating power-conscious keyword spotting solutions, too. The trouble is that the distinction between "things it'd be OK to cut off" (like passive keyword spotting with no recent activity) and "things it wouldn't be OK to cut off" (like a conference call) doesn't exist at the level of the driver that has to inform power management decisions: it just knows that there's a capture stream open and the blanket policy for that is to block sleep transitions. That's true whether you're doing keyword spotting or just capturing into a loopback memory stream that doesn't do anything with a raw `IAudioClient`; the richness of scenario information just isn't accessible at the layers involved.

I know and share the desire for a simple "open this audio capture with a specific power policy that will still sleep" function somewhere -- that unfortunately just doesn't exist in any form today, though, and creating it requires some very deep driver-level breaking changes that are well beyond my realm of expertise to speak to.

Listening for an appropriate set of events and responding to it with the code is for sure the recommended approach and it's what a lot of OS components themselves do--which is likely why has often seemed like Cortana doesn't suffer from this problem (the underlying platform pieces, like SpeechRuntime in the pre-modern Cortana app, do the listen/respond).

I think I wasn't clear about what I meant with the "40 minutes." There wouldn't be an extra 40 minutes that's not responsive--rather, it'd stop responding just as soon as the screen turns off (but no sooner). 

To rephrase my understanding:
* User settings: turn the screen off after 10 minutes, sleep after 30 minutes
* Sequence of events:
  1. Keyword spotter is running, active capture blocks sleep but not screen off
  2. After 10 minutes, the screen turns off and a "display off" signal is sent
  3. In response, the app stops the keyword spotter and active audio capture -- the screen is off and the keyword spotter isn't running
  4. Assuming nothing else is blocking, either 30 or 20 (30 - 10) minutes later (one of them), the computer will go to sleep
  5. Once the screen turns back on (either with or without an intervening sleep), a "display on" signal is sent
  6. The app then immediately resumes its keyword spotting and active audio capture -- as soon as the screen is back on, so is the keyword spotter

There are also a variety of other signals that can be listened for (power source change, away mode, etc.) that can be explored to get a desired effect: https://docs.microsoft.com/en-us/windows/win32/power/power-setting-guids . I *do not* believe there's a "just in time" event available because the capture activity in question is actually suppressing those signals (by design).

As for hardware keyword spotting: this isn't in-built into microphones, but rather the DSP/chipset from a silicon provider (e.g. Intel, AMD). In essence, a lightweight, offloaded algorithm is "baked in" to the chip, which allows early-stage capture and preliminary activation to occur without the CPU involved--which means low power states are possible. Given the coordination involved with custom drivers, this has just been a Cortana and (more recently) Alexa capability so far. There's some exciting industry progress on field-programmable HWKWS capabilities that I expect will make this way more accessible to developers in the coming product cycles, but it'll require new hardware and it's still an indefinite ways off.

**Comment by assitive** [comment]
Thanks for the detailed suggestion. We found that if we set screen off time to "Never" in the power setting, then it fails to sleep automatically. We are still hoping to have a solution working for all the cases. But this might be good for now.

**Comment by pankopon** [comment]
Closing the issue as the questions have been very thoroughly answered.
